lucene/analysis/common/src/java/org/apache/lucene/analysis/util/AnalysisSPILoader.java

```diff
@@ -83,17 +74,22 @@ public synchronized void reload(ClassLoader classloader) {
     final SPIClassIterator<S> loader = SPIClassIterator.get(clazz, classloader);
     while (loader.hasNext()) {
       final Class<? extends S> service = loader.next();
-      final String clazzName = service.getSimpleName();
       String name = null;
-      for (String suffix : suffixes) {
-        if (clazzName.endsWith(suffix)) {
-          name = clazzName.substring(0, clazzName.length() - suffix.length()).toLowerCase(Locale.ROOT);
-          break;
+      Throwable cause = null;
+      try {
+        // Lookup "NAME" field with appropriate modifiers. Also it must be a String field.
+        final Field field = service.getField("NAME");
+        int modifier = field.getModifiers();
+        if (Modifier.isPublic(modifier) && Modifier.isStatic(modifier) && Modifier.isFinal(modifier) &&
```

Comment:
Actually, you can leave the check inside to make it more explicit, I just wanted to mention this: <https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Class.html#getField(java.lang.String)>: "Returns a Field object that reflects the specified public member field of the class or interface represented by this Class object"

Reply:
Thanks for the comment,
I noticed another method `Class#getDeclaredField(String)` which does not traverse superclasses.
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Class.html#getDeclaredField(java.lang.String)
Should we use this instead of `getField(String)` ?

Reply:
I was thinking last night about this, too! My answer: YES! This allows us to have a full check that everything is correct.

Reply:
But I have no idea about the consequences with security manager... So maybe check this.

Reply:
I have a better idea: Use `getField()` and remove the `isPublic()` check. But do `Objects.equals(field.getDeclaringClass(),  service)` as safety check. That has no security manager problems and does the check, too.

Reply:
The general problem of the `getDeclaredField()` for SPI is: The SPI interfaces may be loaded from another classloader (e.g., in Solr) - becaus ethat's the whole idea about SPI!

Reply:
OK, I will change the safety check in that manner.